diff --git a/src/templates/gderived-defs b/src/templates/gderived-defs
index be232c1e0..fe9a383df 100644
--- a/src/templates/gderived-defs
+++ b/src/templates/gderived-defs
@@ -156,13 +156,19 @@ define: (ClassBodyDeclarations)ctr_userdict
 define: (ClassBodyDeclarations)toString
   public String toString() {
     PyType self_type = getType();
-    PyObject impl = self_type.lookup("__repr__");
+    PyObject impl = self_type.lookup("toString");
     if (impl != null) {
-      PyObject res = impl.__get__(this,self_type).__call__();
-      if (!(res instanceof PyString))
+      Object res = impl.__get__(this,self_type).__call__();
+      if (res instanceof PyString) {
+        return ((PyString)res).toString();
+      } else if (res instanceof String){
+        return (String)res;
+      } else if (res instanceof PyObject){
         throw Py.TypeError("__repr__ returned non-string (type "
-                           + res.getType().fastGetName() + ")");
-      return ((PyString)res).toString();
+                           + ((PyObject)res).getType().fastGetName() + ")");
+      }
+      throw Py.TypeError("__repr__ returned non-string (type "
+                         + getClass().getCanonicalName() +")");
     }
     return super.toString();
   }
